### PR TITLE
CLDC-2068 request, validate UPRN and retrieve address

### PR DESCRIPTION
**Pull request by ghost**
new features/behaviours:
- `Question#plain_label`: so that their header is displayed unformatted
- `Question#notification_banner`: to display the notification component with dynamic content
- `Page#skip_text`: To dynamically display alternative text on skip button
- `Page#skip_href`: To dynamically set alternative href on skip button
- Local Authority Known question only asked if year < 2023

New questions:
- UPRN known
- UPRN
- UPRN confirmation
- Full address with Address line 1, Address line 2, County, Town or City, Postcode

new DB fields: 
- uprn
- uprn_known
- uprn_confirmed
- address_line1
- address_line2
- town_or_city
- county

Jira: https://digital.dclg.gov.uk/jira/browse/CLDC-2068
Figma: https://www.figma.com/file/zeKXCOjR1DuXXIXpWHaDwS/CORE-Dev-Handover?node-id=3875%3A82799

For testing purposes:
- valid UPRN: 100023336963
- invalid UPRN: 100023336961

### CODE REVIEW

File: app/models/form/question.rb [hunk 1]
```diff
@@ -56,6 +57,8 @@ def answer_label(log, user = nil)
     inferred_answer_value(log) || answer_label
   end
 
+  def notification_banner(_log = nil); end
```
Comment (ghost) [inline app/models/form/question.rb:61]: needed to display this component
Screenshot 2023-03-03 at 14 52 10

File: app/models/form/sales/pages/uprn.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+      Form::Sales::Questions::Uprn.new(nil, nil, self),
+    ]
+  end
+
+  def routed_to?(log, _current_user)
+    log.uprn_known == 1
+  end
+
+  def skip_text
+    "Enter address instead"
```
Comment (ghost) [inline app/models/form/sales/pages/uprn.rb:18]: for custom skip button copy
Screenshot 2023-03-06 at 14 33 20

File: app/helpers/question_view_helper.rb [hunk 3]
```diff
@@ -8,14 +8,16 @@ def caption(caption_text, page_header, conditional)
...
+      size: label_size(page_header, conditional, question),
       tag: label_tag(page_header, conditional),
     }
   end
 
 private
 
-  def label_size(page_header, conditional)
+  def label_size(page_header, conditional, question)
+    return if question.plain_label.present?
```
Comment (ghost) [inline app/helpers/question_view_helper.rb:19]: `nil` value here is needed for "plaintext" input labels
Screenshot 2023-03-06 at 14 55 23

File: app/services/uprn_client.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  def initialize(uprn)
+    @uprn = uprn
+  end
+
+  def call
+    unless response.is_a?(Net::HTTPSuccess) && result.present?
+      # TODO: verify error copy
+      @error = "UPRN not recognised. Check this, or enter address"
+    end
+  rescue JSON::ParserError
```
Comment (jamesrose) [inline app/services/uprn_client.rb:18]: Perhaps rescue from a timeout?

Comment (ghost) [reply app/services/uprn_client.rb:18]: Discovered this [method](https://apidock.com/ruby/v2_5_5/Net/HTTP/max_retries%3D) that can be used to retry automatically

File: app/services/uprn_client.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    @result ||= JSON.parse(response.body).dig("results", 0, "DPA")
+  end
+
+private
+
+  def http_client
+    client = Net::HTTP.new(ADDRESS, 443)
+    client.use_ssl = true
+    client.verify_mode = OpenSSL::SSL::VERIFY_PEER
+    client.max_retries = 3
```
Comment (jamesrose) [inline app/services/uprn_client.rb:32]: Can we lower the timeout here? The default is 60 seconds (x3) so will be terminated. Suggest <10 seconds, max.

File: app/controllers/form_controller.rb [hunk 6]
```diff
@@ -160,7 +160,7 @@ def mandatory_questions_with_no_response(responses_for_page)
   end
 
   def question_is_required?(question)
-    @log.class::OPTIONAL_FIELDS.exclude?(question.id) && required_questions.include?(question.id)
+    @log.optional_fields.exclude?(question.id) && required_questions.include?(question.id)
```
Comment (ghost) [inline app/controllers/form_controller.rb:163]: this is a bugfix, we weren't including dynamic optional fields

File: app/models/form/sales/pages/uprn.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  end
+
+  def skip_href(log = nil)
+    return unless log
+
+    "/#{log.model_name.param_key.dasherize}s/#{log.id}/address"
```
Comment (jamesrose) [inline app/models/form/sales/pages/uprn.rb:24]: Would you prefer not to use a route helper here?

**Review by jamesrose — APPROVED** [review]

**Pull request merged**